File: frontend/src/types/ops.ts
export type OpsCatalogItem = {
  controlTypeId: string;
  parameterId: string;
  controlTypeName: string;
  parameterName: string;
  metricKey: string;
  displayName: string;
  module: string;
  moduleLabel: string;
  area: string;
  valueType: string;
  unit: string;
  canonicalMetric?: string | null;
  covered: boolean;
  latestSampleAtUtc?: string | null;
  latestSampleAtLocal?: string | null;
};

export type OpsCatalogPayload = {
  version: string;
  source: string;
  systemId: string;
  provider: string;
  coverage: {
    total: number;
    covered: number;
    coverageRate: number;
    gatePassed: boolean;
  };
  items: OpsCatalogItem[];
};

export type OpsMetricValue = {
  metricKey: string;
  displayName: string;
  value: number;
  unit: string;
  valueType: string;
  module: string;
  moduleLabel: string;
  area: string;
  recordedAtUtc: string;
  recordedAtLocal: string;
};

export type OpsZoneSnapshot = {
  zone: string;
  latestSampleAtUtc?: string | null;
  latestSampleAtLocal?: string | null;
  metrics: OpsMetricValue[];
};

export type OpsModuleSnapshot = {
  module: string;
  moduleLabel: string;
  zoneCount: number;
  metricCount: number;
};

export type OpsLivePayload = {
  meta: {
    provider: string;
    lookbackHours: number;
    pageRefreshedAt: string;
    latestSampleAtUtc?: string | null;
    latestSampleAtLocal?: string | null;
    freshnessStatus: "FRESH" | "WARNING" | "STALE";
    warningMessage?: string | null;
    timezone: string;
    storageTimezone: string;
  };
  zones: OpsZoneSnapshot[];
  modules: OpsModuleSnapshot[];
};

export type OpsTrendPoint = {
  metricKey: string;
  displayName: string;
  module: string;
  moduleLabel: string;
  bucketStartUtc: string;
  bucketStartLocal: string;
  avg: number;
  min: number;
  max: number;
  count: number;
};

export type OpsTrendsPayload = {
  provider: string;
  startUtc: string;
  endUtc: string;
  zone?: string | null;
  points: OpsTrendPoint[];
};

File: backend/app/schemas/ops.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


OpsValueType = Literal["numeric", "status", "percentage"]


class MetricCatalogItem(BaseModel):
    controlTypeId: str
    parameterId: str
    controlTypeName: str
    parameterName: str
    metricKey: str
    displayName: str
    module: str
    moduleLabel: str
    area: str
    valueType: str
    unit: str
    canonicalMetric: str | None = None
    covered: bool
    latestSampleAtUtc: datetime | None = None
    latestSampleAtLocal: str | None = None


class OpsCatalogCoverage(BaseModel):
    total: int
    covered: int
    coverageRate: float
    gatePassed: bool


class OpsCatalogResponse(BaseModel):
    version: str
    source: str
    systemId: str
    provider: str
    coverage: OpsCatalogCoverage
    items: list[MetricCatalogItem] = Field(default_factory=list)


class OpsMetricValue(BaseModel):
    metricKey: str
    displayName: str
    value: float
    unit: str
    valueType: str
    module: str
    moduleLabel: str
    area: str
    recordedAtUtc: datetime
    recordedAtLocal: str


class OpsZoneSnapshot(BaseModel):
    zone: str
    latestSampleAtUtc: datetime | None = None
    latestSampleAtLocal: str | None = None
    metrics: list[OpsMetricValue] = Field(default_factory=list)


class OpsModuleSnapshot(BaseModel):
    module: str
    moduleLabel: str
    zoneCount: int
    metricCount: int


class OpsLiveMeta(BaseModel):
    provider: str
    lookbackHours: int
    pageRefreshedAt: datetime
    latestSampleAtUtc: datetime | None = None
    latestSampleAtLocal: str | None = None
    freshnessStatus: Literal["FRESH", "WARNING", "STALE"]
    warningMessage: str | None = None
    timezone: str = "Asia/Shanghai"
    storageTimezone: str = "UTC"


class OpsLiveResponse(BaseModel):
    meta: OpsLiveMeta
    zones: list[OpsZoneSnapshot] = Field(default_factory=list)
    modules: list[OpsModuleSnapshot] = Field(default_factory=list)


class OpsTrendPoint(BaseModel):
    metricKey: str
    displayName: str
    module: str
    moduleLabel: str
    bucketStartUtc: datetime
    bucketStartLocal: str
    avg: float
    min: float
    max: float
    count: int


class OpsTrendsResponse(BaseModel):
    provider: str
    startUtc: datetime
    endUtc: datetime
    zone: str | None = None
    points: list[OpsTrendPoint] = Field(default_factory=list)
